### ISSUE TITLE
Support of SMOTE with cross-validation

### ISSUE THREAD

**Issue by petrosDemetrakopoulos** [microsoft/FLAML#1200]
Hello, congrats for the project. It is purely awesome.

I have a small enhancement to suggest:
Sometimes we have to deal with imbalanced datasets and we employ "Synthetic Minority Oversampling Techniques" (SMOTE).
However, oversampling needs to **only be done in the training set** and not in the validation/test sets.
Thus, in a cross-validation scenario, at each split SMOTE needs to be done only on the train split and not on validation split. 

This functionality is not currently supported by FLAML.
I would suggest the SMOTE over-sampler object be passed in the automl_settings and then used at each split.

The imbalanced-learn package achieves the same functionality using custom sklearn-like Pipelines.

**Comment by immu4989** [comment]
Picking this up — happy to take it on. Before I write any code, want to align with maintainers on the API shape since this is the first per fold preprocessing hook to land in evaluate_model_CV.

**Proposed API**
```
from imblearn.over_sampling import SMOTE

automl.fit(
    X_train, y_train,
    resampler=SMOTE(random_state=42),   # any object with .fit_resample(X, y) -> (X, y)
    ...                                  # all other settings as before
)
```
Applied inside evaluate_model_CV between the per-fold split and the estimator fit (around flaml/automl/task/generic_task.py:1234), so the resampler sees the train partition only — never the validation rows.

**Open design questions**

1. Contract: narrow to imbalanced-learn's BaseSampler protocol (fit_resample(X, y) -> (X, y)), or accept any callable with the same signature? Narrow protocol is safer for input validation; callable is more flexible.

2. Optional dependency: add imbalanced-learn as an extra (pip install flaml[imbalanced]) and runtime-check the import, mirroring how catboost/xgboost are handled today?

3. sample_weight interaction: resampling breaks the 1 to 1 row alignment with user-supplied sample_weight. Three options:

    (a) Reject the combination with a clear error
    (b) Drop sample_weight after resampling and warn
    (c) Require the user's resampler to also resample the weights (less common in practice)
    I'd default to (a) for v1 — explicit, no silent surprise.
    
4. Reproducibility: any preference on whether FLAML should auto-seed the resampler from the random_seed config (matching the recently-landed audit work in #1540), or strictly defer to the user's own random_state on the sampler object?

5. Scope of v1: stick to per fold train time resampling only. Test-time prediction stays unchanged (predict on the original feature space). Resampler ensemble methods like BalancedBaggingClassifier that need prediction-time logic are out of scope for v1 — those should use their own sklearn-compatible estimator path.

**Test plan**

- New test_imbalanced.py (or extend test_classification.py) covering:
- Smoke test on a synthetic imbalanced binary dataset (make_classification(weights=[0.95, 0.05]))
- Verify the resampler is not applied to validation folds (assert val-fold class ratio unchanged)
- Verify reproducibility (same random_seed → same selected config)
- Verify the rejection error in (3a) above

Happy to adjust any of this — flagging now since the API shape is harder to change later. I'll start implementation once we have alignment on (1)-(5).

**Comment by immu4989** [comment]
@thinkall — bumping this in case the comment above got missed. Happy to proceed with the following defaults and open a draft PR for review if you'd prefer to discuss against concrete code:

- (Q1, contract) narrow to imbalanced-learn's BaseSampler protocol (fit_resample(X, y) -> (X, y))
- (Q2, optional dep) add imbalanced-learn as an extra (pip install flaml[imbalanced]) with a runtime import check
- (Q3, sample_weight + resampler) option (a) — reject the combination with a clear error in v1
- (Q4, reproducibility) auto-seed the resampler from the FLAML random_seed config, matching the recently landed audit (#1540)
- (Q5, scope) train-time per-fold resampling only; prediction stays on the original feature space

**Comment by thinkall** [comment]
Thank you so much for bringing this up. @immu4989 , this will be a big change, please allow me some time to think it through . I've considered this before, but currently recommending users to apply SMOTE before passing data to FLAML. How much performance gain on the final prediction can we have vs how many extra training budget is needed? I'd prefer some benchmarks on this. WDYT? Thanks.

**Comment by immu4989** [comment]
Sounds good , fair ask. Let me put together a small benchmark across a couple of representative imbalanced classification datasets (one synthetic, one real, varied imbalance ratios), comparing:

- A: No SMOTE (baseline)
- B: SMOTE pre applied to the full training set before AutoML.fit() (the workaround you currently recommend)
- C: per fold SMOTE inside evaluate_model_CV (this proposal)

Reporting minority class F1 / recall / PR-AUC, plus wall time and total trials, all reproduced over a few seeds. Will post results here before any code lands. Aiming for next week

**Comment by immu4989** [comment]
# SMOTE-with-CV benchmark for FLAML #1200

## Setup

- **Datasets** — two synthetic binary classification datasets (`make_classification`, n=1500 each):
  - `synthetic_7pct` — 7% minority class
  - `synthetic_15pct` — 15% minority class
- **Outer split** — 80/20 stratified holdout per (dataset, seed).
- **Seeds** — 0, 1, 2.
- **FLAML config** — `estimator_list=['lgbm']`, `eval_method='cv'`, `n_splits=5`, `max_iter=12`, `metric='f1'`, `n_jobs=1`, `retrain_full=True`.
- **Conditions:**
  - **A** — `automl.fit(X, y)` — no SMOTE
  - **B** — `X', y' = SMOTE().fit_resample(X, y); automl.fit(X', y')` — workaround currently recommended
  - **C** — per-fold SMOTE inside `evaluate_model_CV` (monkey-patched `get_val_loss`), then `retrain_full=True` on the **raw** training set
  - **C2** — same per-fold CV as C, but the final retrain also uses `SMOTE(X, y)` so the search-target and the deployed model agree on class distribution
- **Metrics** — minority-class precision / recall / F1, PR-AUC, wall-clock seconds in `automl.fit`.

Reproducibility: `bench_smote/run_bench.py` + `results.json` in the gist.

## Headline results (mean across 3 seeds)

| Dataset | Condition | Precision | Recall | F1 | PR-AUC | Fit secs |
|---|---|---:|---:|---:|---:|---:|
| 15% min | A no_smote | 0.836 | 0.521 | 0.636 | 0.720 | 0.24 |
| 15% min | B smote_full | 0.593 | 0.706 | **0.639** | **0.742** | 0.31 |
| 15% min | C per_fold | 0.786 | 0.520 | 0.620 | 0.664 | 0.40 |
| 15% min | C2 per_fold + retrain_resampled | 0.622 | 0.662 | 0.638 | 0.711 | 0.51 |
| 7% min  | A no_smote | 0.775 | 0.392 | **0.519** | 0.617 | 0.23 |
| 7% min  | B smote_full | 0.435 | 0.594 | 0.498 | 0.616 | 0.44 |
| 7% min  | C per_fold | 0.681 | 0.293 | 0.403 | 0.520 | 0.43 |
| 7% min  | C2 per_fold + retrain_resampled | 0.441 | 0.678 | 0.513 | 0.594 | 0.49 |

## Findings

1. **B vs A** — pre-applied SMOTE gives a small PR-AUC bump on moderate imbalance (15% dataset: 0.742 vs 0.720) and roughly ties on F1. On the more imbalanced 7% dataset there's no advantage on either metric — A wins F1, B wins recall, neither wins PR-AUC.
2. **C is the weakest of all four** on both datasets. Diagnosis: the hyperparameter search is conditioned on SMOTE-expanded fold-training distributions, but `retrain_full=True` fits the final deployed model on the raw distribution. The hyperparameters and the deployed model see different class ratios.
3. **C2 closes most of the gap to B** by resampling the final retrain as well, so the search target and the deployed model agree. C2 ≈ B on F1 across both datasets; C2 has a slight PR-AUC edge over A on the 7% dataset, but no clear edge over B.
4. **Cost** — C2 is ~40–60 % slower than B (SMOTE runs 6× — once per fold plus once for retrain — vs 1× in B). For the gains observed, that's an unfavorable cost/benefit on these datasets.

## Interpretation

For binary classification with LightGBM at moderate-to-heavy imbalance, the per-fold-SMOTE integration proposed in #1200 does **not** materially outperform the simpler workaround you currently recommend. The current workaround **B** is competitive with the more architecturally correct **C2** on the metrics that matter (F1, PR-AUC), at roughly half the wall-clock cost.

The story is *not* that per-fold resampling is wrong — it's that on these problems, LightGBM's own class-imbalance handling and FLAML's hyperparameter search jointly compensate for the leakage in B such that the cleaner C2 path doesn't show measurable lift. The benefit may be more visible:
- on harder estimators that don't handle imbalance natively (LinearSVC, sklearn `LogisticRegression`);
- with larger hyperparameter search budgets (`max_iter ≫ 12`);
- with real datasets where the leakage in B leads to overconfident validation estimates that mislead the search.

## Recommended next step

Three options, ordered from smallest scope to largest:

1. **Park the feature.** Document the current pre-applied-SMOTE workaround in the FLAML docs (one example notebook) and close #1200 as "use the existing workaround." Cheapest, fastest, honest given these numbers.
2. **Re-run the benchmark with more diverse settings** before deciding — add a real imbalanced dataset (e.g. `credit-card-fraud`), add a non-LGBM estimator (`lrl1` / `lrl2`), bump `max_iter` to 50+, and re-evaluate. If C2 shows clear lift in any of those slices, ship the feature; otherwise close as in (1).
3. **Ship a minimal `resampler=` kwarg anyway** as a convenience for users who currently roll their own per-fold pipelines (it's clean code and matches imblearn idioms), but be explicit in the docs that it isn't expected to materially improve quality on most problems vs the pre-applied workaround.

My vote is **(2)** — the benchmark above is enough to say "no clear win in the easy case" but not enough to say "no win ever." I'm happy to run the expanded sweep if you'd like a second pass before deciding. If you'd rather skip straight to (1) or (3) based on these numbers, also fine.

## Caveats

- Single estimator (`lgbm`) — LightGBM has good built-in imbalance handling, which biases the comparison against SMOTE.
- Synthetic data only — `make_classification`'s minority cluster geometry may be easier than real-world imbalanced data.
- Default SMOTE — no `BorderlineSMOTE`, `ADASYN`, `SMOTENC` variants.
- Small search budget (`max_iter=12`) — a bigger budget might let C2 explore more of the SMOTE-friendly hyperparameter region.
- 3 seeds — variance is non-trivial (e.g. C on 7%-seed-1 had F1=0.258 vs F1=0.514 on seed-2); a larger seed pool would tighten the means.

**Comment by immu4989** [comment]
@thinkall , friendly bump. The benchmark above gives us A vs B vs C/C2 numbers on two synthetic imbalance levels, with the headline that per fold SMOTE inside FLAML's CV doesn't materially outperform the pre-applied SMOTE workaround you currently recommend (the latter is in fact roughly tied with C2 on F1 and PR-AUC, at about half the wall clock cost).

Three concrete next steps were proposed there. To repeat them briefly:

1. **Park the feature** — document the pre-applied SMOTE workflow and close as "use the existing workaround." The new [Production-Deployment guide](https://github.com/microsoft/FLAML/blob/main/website/docs/Use-Cases/Production-Deployment.md#5-sample-weights-and-cost-sensitive-learning) (merged in #1562) already points users to this pattern.
2. **Run an expanded benchmark sweep** — add a real imbalanced dataset, a non-LGBM estimator (e.g. `lrl1`), larger `max_iter`, before deciding. I'm happy to run this if you want a second pass.
3. **Ship a minimal `resampler=` kwarg anyway** as a convenience for users with their own per-fold pipelines, with docs caveats that it isn't expected to materially improve metrics on most problems.

No urgency from my end, just want to make sure the thread doesn't get lost. Let me know which direction you'd like and I'll act on it, or close it out if you'd prefer.

**Comment by jeet2005** [comment]
Hi — I'd like to take this on. I work with sklearn pipelines and imbalanced datasets regularly (built an AutoML platform with a 56-algorithm registry across sklearn/XGBoost/LightGBM/CatBoost), and I went through evaluate_model_CV in flaml/automl/task/generic_task.py to scope this properly before proposing an approach.
Where the gap actually is:
In evaluate_model_CV, each fold currently does:
pythonX_train = X_train_split.iloc[train_index]
X_val   = X_train_split.iloc[val_index]
y_train, y_val = y_train_split[train_index], y_train_split[val_index]
and passes X_train, y_train straight into get_val_loss → estimator.fit(). There's no hook to transform the train fold before fitting, so today the only way to do SMOTE-with-CV correctly is to monkeypatch this loop yourself.
Proposed fix (3 parts):

Add a resampler kwarg to AutoML.__init__/fit(), following the exact precedent already set by auto_augment (stored as self._resampler, threaded through to the search state the same way). Accepts any object with a fit_resample(X, y) method — i.e. imblearn's SMOTE, RandomOverSampler, etc., no new abstraction needed.
Apply it inside evaluate_model_CV, after the fold split, before get_val_loss:

python   if resampler is not None:
       X_train, y_train = resampler.fit_resample(X_train, y_train)
X_val/y_val stay untouched — which is the whole point of the issue (no oversampling leakage into validation).

Mirror the same one-line hook in evaluate_model_holdout, so the fix covers both eval_method="cv" and "holdout", not just CV.

Why not wrap the estimator in an imblearn.Pipeline instead (the original suggestion)? FLAML's estimators take custom fit_kwargs per learner — sample_weight, groups, early-stopping callbacks for LightGBM/XGBoost — and a generic imblearn Pipeline wrapper would silently break those kwarg paths. Doing the resample explicitly in the eval functions is more surgical and doesn't touch the estimator interface.
Test I'd add: parametrized test asserting (a) the validation fold's class distribution is unchanged from the original imbalanced data, and (b) the train fold's minority class count increases post-resample.
Open question for maintainers: should imbalanced-learn be a hard dependency or optional-import like holidays is for time series? I'd lean optional, gated the same way.
Happy to start on this once aligned — let me know if there's a preferred name for the kwarg (resampler vs. something else) or if you'd rather see this live as a custom_hp-style config instead of a new top-level argument.

**Comment by immu4989** [comment]
Thanks for the engagement, @jeet2005.

Two clarifications from my side:

**1. The implementation shape you describe is exactly what my benchmark above already evaluates.** The `resampler` kwarg → per-fold hook in `evaluate_model_CV` is Condition C / C2 in [my benchmark results](https://github.com/microsoft/FLAML/issues/1200#issuecomment-4705108730) (posted 2026-06-15) — implemented via the same insertion point you sketched, monkey-patched on `get_val_loss` rather than the eval function but functionally identical. The headline: on the two synthetic datasets I ran (7% and 15% minority), per-fold SMOTE did **not** materially outperform the pre-applied SMOTE workaround @thinkall currently recommends, and the C variant that doesn't also resample the final retrain was the worst of the four. So the open question right now isn't *how* to implement it — it's *whether* to ship it given those numbers, which is what the three options at the bottom of my benchmark comment are asking @thinkall to choose between.

**2. I've taken this thread end-to-end and would prefer not to fork the implementation.** Claim posted 2026-05-23, API proposal 2026-05-28, benchmark delivered 2026-06-15, ping today. I'm committed to running the expanded sweep if @thinkall picks option (2) above, or to shipping the convenience kwarg if (3) — please don't open a competing PR while the maintainer decision is pending.

The most useful contribution right now would be a benchmark dataset / estimator combination you think would change the picture from the synthetic-LGBM slice — if you have one in mind from your imbalanced-data work, posting it here would be a real help in pushing this to a decision.

**Comment by thinkall** [comment]
Thank you so much for the detailed benchmark, @immu4989 ! Many thanks to the discussion, @jeet2005 . 

@immu4989, I'd prefer option (3) as internal SMOTE support can definitely help some users, but I hope to keep it optional and off by default. Does it make sense to you? Thanks.

**Comment by immu4989** [comment]
Thanks @thinkall. Option (3), off by default, makes sense to me.

Locked-in design from the discussion above:

- New `resampler=` kwarg on `AutoML.fit`, defaulting to `None` (no behavior change for existing users).
- Accepts any object exposing `fit_resample(X, y) -> (X, y)` — matches imbalanced-learn's `BaseSampler` protocol.
- Applied inside `evaluate_model_CV` and `evaluate_model_holdout` on the per-fold training partition only. Validation partitions stay at the raw class distribution.
- imbalanced-learn stays an **optional** dependency (runtime import inside `fit_resample` via the user-supplied object; FLAML itself doesn't import it). Users install it themselves if they want to use it.
- If `sample_weight` is in `fit_kwargs` alongside a `resampler`, raise a clear `ValueError` at fit-time — resampling breaks the 1-to-1 row alignment with weights, and silently dropping either would be a footgun. Q3(a) from my earlier proposal.
- Docs note in the [Production-Deployment guide](https://github.com/microsoft/FLAML/blob/main/website/docs/Use-Cases/Production-Deployment.md#5-sample-weights-and-cost-sensitive-learning) will point users to this kwarg with a caveat that the pre-applied SMOTE workaround remains a valid choice — the benchmark above shows the two are roughly equivalent on most problems, so neither is strictly better.

Will open a PR shortly with the kwarg, the per-fold hook in both CV and holdout paths, and a parametric test covering (a) validation fold's class ratio is unchanged, (b) train fold's minority count increased, (c) reject `sample_weight` + `resampler` combination.
